frame/ethereum/src/lib.rs

```diff
@@ -563,14 +668,73 @@ impl<T: Config> Pallet<T> {
 	/// Execute an Ethereum transaction.
 	pub fn execute(
 		from: H160,
-		input: Vec<u8>,
-		value: U256,
-		gas_limit: U256,
-		gas_price: Option<U256>,
-		nonce: Option<U256>,
-		action: TransactionAction,
+		transaction: &Transaction,
 		config: Option<evm::Config>,
 	) -> Result<(Option<H160>, Option<H160>, CallOrCreateInfo), DispatchError> {
+		let (
+			input,
+			value,
+			gas_limit,
+			max_fee_per_gas,
+			max_priority_fee_per_gas,
+			nonce,
+			action,
+			access_list,
+		) = {
+			match transaction {
+				// max_fee_per_gas and max_priority_fee_per_gas in legacy and 2930 transactions is
+				// the provided gas_price.
+				Transaction::Legacy(t) => {
+					let base_fee = T::FeeCalculator::min_gas_price();
+					(
+						t.input.clone(),
+						t.value,
+						t.gas_limit,
+						Some(base_fee),
+						Some(t.gas_price.saturating_sub(base_fee)),
+						Some(t.nonce),
+						t.action,
+						Vec::new(),
+					)
+				}
+				Transaction::EIP2930(t) => {
+					let base_fee = T::FeeCalculator::min_gas_price();
+					let access_list: Vec<(H160, Vec<H256>)> = t
+						.access_list
+						.iter()
+						.map(|item| (item.address, item.slots.clone()))
+						.collect();
+					(
+						t.input.clone(),
+						t.value,
+						t.gas_limit,
+						Some(base_fee),
+						Some(t.gas_price.saturating_sub(base_fee)),
```

Comment:
If for whatever reason the pre-validation is skipped, here could be the case that the `t.gas_price `is less than `base fee`. It's better to `checked_sub` instead `saturating_sub`, and reject the execution in fails.